naga/src/front/wgsl/parse/mod.rs

```diff
@@ -2433,4 +2454,43 @@ impl Parser {
         }
         Ok(brace_nesting_level + 1)
     }
+
+    fn diagnostic_control<'a>(&self, lexer: &mut Lexer<'a>) -> Result<(), Error<'a>> {
+        // TODO(maybe): self.push_rule_span(Rule::???, lexer)
```

Comment:
@gfx-rs/naga: It's unclear to me when to use `Rule` spans. Should they be used for this new functionality? 🤔

Reply:
I have always been very unsure what sorts of bugs the WGSL front end's `Rule` stack makes it easy to detect or fix. In a recursive descent parser, the parser's own Rust call stack indicates what we were in the midst of parsing, and the `Rule` stack is either identical to that, or misleading. 

It does seem like the code uses `push_rule_span` and `pop_rule_span` to build spans that cover non-trivial subtrees. But you can do that just as well by calling `Lexer::span_from` yourself; that's something you see in the code as well.

My personal preference would be to just use `Lexer::span_from` directly. Logically, it should be easier to debug span problems when there are fewer "players" influencing how they're produced.